[Merged by Bors] - feat(Algebra/GCDMonoid): add some theorems related to divisibility and `lcm` for both `GCDMonoid` and `Nat`

ShreckYe opened the pull request:
While working on formalizing some math olympiad problems, I found that the theorem `dvd_lcm : p ∣ lcm a b ↔ p ∣ a ∨ p ∣ b` is missing in Mathlib. So I added this along with some other related theorems. These theorems are simlilar to those related to divisibility and multiplication (`dvd_mul` as compared to `dvd_lcm` for example), as commented in the doc strings.

ShreckYe committed:
Add some theorems related to divisibility and `lcm` for `GCDMonoid`

ShreckYe committed:
Remove some redundant code which is already commented out

ShreckYe committed:
Adjust the doc strings as they are not complete sentences

ShreckYe committed:
Adapt the added theorems for `Nat.lcm` and `Nat.Prime` too

ShreckYe committed:
More as in commit 0ac00c2e4ac91bc6079253248e25e7da54c32cc4

ShreckYe commented:
As the part of the added code involves both `Prime` and `lcm`, I am not sure which of the scopes/directories `Prime` or `GCDMonoid`/`GCD` is a better place to put these new theorems in. Please help me move the code or let me know if there are better ways to organize them.

ShreckYe committed:
Run `lake exe shake --fix` and keep the changes in files I edited to fix a CI failure

themathqueen commented on Mathlib/Algebra/GCDMonoid/Basic.lean:
```diff
@@ -808,6 +808,60 @@ theorem lcm_eq_of_associated_right [NormalizedGCDMonoid α] {m n : α} (h : Asso
+@[simp] theorem lcm_dvd_mul [GCDMonoid α] (m n : α) : lcm m n ∣ m * n :=
+  lcm_dvd (by simp) (by simp)
+
+/-- adapted from and similar to `dvd_mul_of_dvd_left` -/
```
```suggestion
```
should probably remove these everywhere

themathqueen commented on Mathlib/Algebra/GCDMonoid/Basic.lean:
```diff
@@ -808,6 +808,60 @@ theorem lcm_eq_of_associated_right [NormalizedGCDMonoid α] {m n : α} (h : Asso
+theorem dvd_of_lcm_left_dvd [GCDMonoid α] {a b c : α} (h : lcm a b ∣ c) : b ∣ c :=
+  (dvd_lcm_right a b).trans h
+
+--end Divisibility
```
```suggestion
end Divisibility
```

themathqueen commented on Mathlib/Algebra/GCDMonoid/Basic.lean:
```diff
@@ -808,6 +808,60 @@ theorem lcm_eq_of_associated_right [NormalizedGCDMonoid α] {m n : α} (h : Asso
   dvd_antisymm_of_normalize_eq (normalize_lcm _ _) (normalize_lcm _ _) (lcm_dvd_lcm dvd_rfl h.dvd)
     (lcm_dvd_lcm dvd_rfl h.symm.dvd)
 
+-- TODO consider reorganizing and extracting the variables here
+/-
+section Divisibility
+
+variable [GCDMonoid α] {m n a b c : α}
+ -/
```
and then remove them in each result below

themathqueen commented on Mathlib/Algebra/GCDMonoid/Basic.lean:
```diff
@@ -808,6 +808,60 @@ theorem lcm_eq_of_associated_right [NormalizedGCDMonoid α] {m n : α} (h : Asso
+ -/
+
+-- adapted from `Nat.lcm_dvd_mul` in Lean 4's source
+@[simp] theorem lcm_dvd_mul [GCDMonoid α] (m n : α) : lcm m n ∣ m * n :=
```
with the suggestion above ^
```suggestion
variable (m n) in
@[simp] theorem lcm_dvd_mul : lcm m n ∣ m * n :=
```

themathqueen commented on Mathlib/Algebra/GCDMonoid/Basic.lean:
```diff
@@ -808,6 +808,60 @@ theorem lcm_eq_of_associated_right [NormalizedGCDMonoid α] {m n : α} (h : Asso
+  dvd_or_dvd hp (h.trans (lcm_dvd_mul a b))
+
+/-- adapted from and similar to `Prime.dvd_mul` -/
+theorem dvd_lcm : p ∣ lcm a b ↔ p ∣ a ∨ p ∣ b :=
```
```suggestion
theorem dvd_lcm_iff : p ∣ lcm a b ↔ p ∣ a ∨ p ∣ b :=
```

ShreckYe committed:
Remove some comments and doc strings

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Remove more similar doc strings

ShreckYe committed:
Extract common variables and add a section

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Remove the extracted variables

ShreckYe committed:
Move the `Prime` namespace into the `Divisibility` section too

Ruben-VandeVelde commented on Mathlib/Data/Nat/GCD/Basic.lean:
```diff
@@ -7,6 +7,7 @@ import Batteries.Data.Nat.Gcd
 import Mathlib.Algebra.Group.Nat.Units
 import Mathlib.Algebra.GroupWithZero.Divisibility
 import Mathlib.Algebra.GroupWithZero.Nat
+import Mathlib.Data.Nat.Prime.Defs
```
Please put the results that need this import on a separate file

Ruben-VandeVelde reviewed (approved):
Looks good, but the import increase is a bit heavy for my tastes

ShreckYe committed:
Move the code that needs `import Mathlib.Data.Nat.Prime.Defs` into a separate file as requested

ShreckYe committed:
`lake exe mk_all` and add a copyright string and a module doc-string to fix the CI failure

ShreckYe committed:
Merge branch 'master' into dvd_lcm

ShreckYe committed:
Remove a blank line

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Remove a blank line

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Remove a blank line

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Add a blank line

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Add a blank line

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

ShreckYe committed:
Update the module doc-string

ShreckYe committed:
List the main results in the module doc-string as suggested by @themathqueen using Cursor

Prompt:

```
Say what the main results are in the file in the module doc-string, in a way similar to how they are listed in other files.
```

ShreckYe committed:
Revert a line edited by Cursor

themathqueen commented on Mathlib/Data/Nat/GCD/Prime.lean:
```diff
@@ -0,0 +1,41 @@
+# Lemmas related to `Nat.Prime`
+
+This file contains lemmas related to `Nat.Prime`.
+These lemmas are kept separate from `Data.Nat.GCD.Basic` in order to minimize imports.
```
In that case, write it as:
```suggestion
These lemmas are kept separate from `Mathlib/Data/Nat/GCD/Basic.lean` in order to minimize imports.
```

themathqueen commented on Mathlib/Data/Nat/GCD/Prime.lean:
```diff
@@ -0,0 +1,41 @@
+import Mathlib.Data.Nat.Prime.Defs
+
+/-!
+# Lemmas related to `Nat.Prime`
```
What about `GCD` since it's in the `Nat/GCD/` folder? Or actually, maybe that's just implied. It's fine, just keep it as is, I guess

ShreckYe committed:
Update the module doc-string

Co-authored-by: themathqueen <23701951+themathqueen@users.noreply.github.com>

erdOne reviewed (approved):
Sorry for neglecting this for so long as the assignee. This PR looks basically good to go except for the minor comment above.

Thanks!
maintainer delegate

ShreckYe committed:
Add `lcm` in the module doc-string

Co-authored-by: Monica Omar <23701951+themathqueen@users.noreply.github.com>

ShreckYe commented:
bors r+
